src/rcur.c

```diff
@@ -0,0 +1,254 @@
+#include "rcur.h"
+#include "ccan/utf8/utf8.h"
+#include "cursor.h"
+#include <errno.h>
+#include <string.h>
+
+/* Note that only rcur_pull and rcur_peek_remainder actually access
+ * rcur.  The rest are built on top of them! */
+
+/* Pull @len bytes from rcur, if space available.
+ * Return NULL if not valid.
+ */
+const void *rcur_pull(struct rcur *rcur, size_t len)
+{
+    const void *p = rcur_peek(rcur, len);
+    if (!p) {
+	rcur_fail(rcur);
+	return NULL;
+    }
+    rcur->p += len;
+    return rcur->p - len;
+}
+
+/* Access the remaining bytes.  Returns NULL and *len=0 if invalid. */
+const void *rcur_peek_remainder(const struct rcur *rcur, size_t *len)
+{
+    if (!rcur_valid(rcur)) {
+	*len = 0;
+	return NULL;
+    }
+    *len = rcur->end - rcur->p;
+    return rcur->p;
+}
+
+bool rcur_copy(struct rcur *rcur, void *dst, size_t len)
+{
+    const void *src = rcur_pull(rcur, len);
+    if (!src)
+	return false;
+    memcpy(dst, src, len);
+    return true;
+}
+
+/* Look ahead: returns NULL if @len too long.  Does *not* alter
+ * rcur */
+const void *rcur_peek(const struct rcur *rcur, size_t len)
+{
+    size_t actual_len;
+    const void *p = rcur_peek_remainder(rcur, &actual_len);
+
+    if (len > actual_len)
+	return NULL;
+    return p;
+}
+
+/* All these are based on rcur_pull. */
+bool rcur_skip(struct rcur *rcur, size_t len)
+{
+    return rcur_pull(rcur, len) != NULL;
+}
+
+unsigned char rcur_pull_byte(struct rcur *rcur)
```

Comment:
this is a very hot function (especially when used in `ndb_text_search_key_compare` > `rcur_pull_varint`) where lmdb can call this many times when walking the btree during search queries.

it seems this hot code path now calls (`rcur_pull` > `rcur_peek` > `rcur_peek_remainer`+`memcpy` for a single byte)

when before it was just:

```
static inline int cursor_pull_byte(struct cursor *cursor, unsigned char *c)
{
        if (unlikely(cursor->p >= cursor->end))
                return 0;

        *c = *cursor->p;
        cursor->p++;

        return 1;
}
```

this makes me a bit nervous, but maybe the compiler will optimize this all away. will try to do some search benchmarking and compare